Source/WebCore/dom/InternalObserverForEach.cpp

```diff
@@ -0,0 +1,143 @@
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        auto* context = scriptExecutionContext();
+        ASSERT(context);
+
+        Ref vm = context->globalObject()->vm();
+
+        JSC::Exception* exception = nullptr;
+        {
+            JSC::JSLockHolder lock(vm);
+
+            // The exception is not reported, instead it is forwarded to the
+            // abort signal and promise rejection.
+            // As such, VisitorCallback `[RethrowsException]` and here a
+            // catch scope is declared so the error can be passed to any promise
+            // rejection handlers and the abort signal.
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_callback->handleEvent(value, m_idx++);
+
+            exception = scope.exception();
+            if (UNLIKELY(exception))
+                scope.clearException();
+        }
+
+        if (UNLIKELY(exception)) {
+            auto value = exception->value();
+            m_promise->reject<IDLAny>(value);
+            m_signal->signalAbort(value);
+            return;
+        }
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        m_promise->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+        m_promise->resolve();
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
+    {
+        m_callback->visitJSFunction(visitor);
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor& visitor) const final
+    {
+        m_callback->visitJSFunction(visitor);
+    }
+
+    InternalObserverForEach(ScriptExecutionContext& context, Ref<VisitorCallback> callback, Ref<AbortSignal>& signal, Ref<DeferredPromise>& promise)
+        : InternalObserver(context)
+        , m_callback(callback)
+        , m_signal(signal)
+        , m_promise(promise)
+    {
+    }
+
+    uint64_t m_idx { 0 };
+    Ref<VisitorCallback> m_callback;
+    Ref<AbortSignal> m_signal;
+    Ref<DeferredPromise> m_promise;
+};
+
+void createInternalObserverOperatorForEach(ScriptExecutionContext& context, Ref<Observable> observable, Ref<VisitorCallback> callback, SubscribeOptions options, Ref<DeferredPromise>&& promise)
+{
+    Ref<AbortSignal> signal = AbortSignal::create(&context);
+
+    Vector<Ref<AbortSignal>> signals = { signal };
+    if (UNLIKELY(options.signal))
+        signals.append(*options.signal);
+
+    Ref dependantSignal = AbortSignal::any(context, signals);
+
+    if (UNLIKELY(dependantSignal->aborted()))
+        return promise->reject<IDLAny>(dependantSignal->reason().getValue());
+
+    dependantSignal->addAlgorithm([promise](JSC::JSValue reason) {
+        promise->reject<IDLAny>(reason);
+    });
+
+    auto observer = InternalObserverForEach::create(context, callback, dependantSignal, promise);
```

Comment:
Note for reviewer: Although all our tests pass, I do not feel this line is correct.

We really should have:

```diff
- auto observer = InternalObserverForEach::create(context, callback, dependantSignal, promise);
+ auto observer = InternalObserverForEach::create(context, callback, signal, promise);
```

as in, pass the `signal` as opposed to `dependantSignal` to mirror [the spec](https://wicg.github.io/observable/#dom-observable-foreach) with these _points of interest_:

_"Let visitor callback controller be a new AbortController."_

and then further

_"If an exception E was thrown, then reject p with E, and signal abort visitor callback controller with E."_

So to me that means `abort` on the `signal` we created here, than abort on the `dependantSignal`. **BUT!** I was unable to get the tests to pass... You'll find the `visitor callback rejection microtask ordering` test to fail, if all we do is swap `dependantSignal` here with `signal` — and 🤷🏻 no idea why.

A simple test to async abort the signal, that it rejects the promise — fails.

```js
let o = new Observable(() => {});
let controller = new AbortController()
let prom = o.forEach(() => {}, { signal: controller.signal };

controller.abort('error');

// NOTE: We hang here, neither reject _nor_ resolve.
try { await prom; } catch (e) {
   assert(e, 'error');
}
```

Through many trial and errors, I found that:

```cpp
dependantSignal->addAlgorithm([promise](JSC::JSValue reason) {
  promise->reject<IDLAny>(reason);
});
```

would never-ever fire. But would if we passed `dependantSignal` down into `InternalObserverForEach`. Which to me feel's wrong.

So, does `AbortSignal::any` do something with how the wiring works, so that if a dependent signal aborts, the "result signal" does not run its algorithms? Or maybe this just C ownership things?

Reply:
I did some further investigation, and Im almost fully convinced the wpt test that were failing — are actually wrong.

Based on [the change to the subscriber spec](https://github.com/WICG/observable/pull/154), and implementation change in [Chrome](https://chromium-review.googlesource.com/c/chromium/src/+/5676226). Which updated the `toArray()` tests, but not the `forEach`.

I decided to remove the failing test, and replace it with a clone of the `toArray()` tests + fix the implementation to now followSignal.

I think however though, we should update the spec to remove all aspects of dependent signals with relation to these promise returning operators — and to instead mirror that of `subscribe`.

Reply:
I did some further investigation, and I am almost fully convinced the spec is _wrong_. We do not want a `dependantSignal`. We are at odds here with the DOM spec though, where the desired behaviour we want, is not specified anymore (see https://github.com/whatwg/dom/pull/1152). But have opted to use it anyway, as that is what the wpt tests assume the timing to be.

As for the above comment, the reason the tests was failing with the dependent signal approach, was largely due to dependent signals not forwarding one signals abort status to all other sibling signals.

You can try this yourself with:

```js
const a = new AbortController();
const b = new AbortController();
const c = AbortSignal.any([a.signal, b.signal]);
b.abort();

a.signal.aborted; // false <~ aborting b does not abort a
b.signal.aborted; // true
c.aborted; // true
```

Which in our case would be that `a` is our internal abort signal, `b` the `SubscribeOptions.signal`, and `c` the [deprived signal point 3](https://wicg.github.io/observable/#dom-observable-foreach) tells us to create.

Further; I ended up copying the tests from `toArray` as to me they were much easier to understand the ordering, than microtask assertions `forEach` was doing. We want all promise returning operators to have a consistent and predictable abort timing order — so  I will work with @dominiccooney and @domfarolino on making those spec changes.

Given this feature is still behind a feature flag, and the wpt tests pass on both Chrome and _now us_ — I am inclined to not let that spec change halt landing this PR.

Reply:
I did some further investigation, and I am almost fully convinced that either the spec is _wrong_ — or wpt tests are wrong. The desired behaviour we want (where handlers all fire as wpt wants), is not specified anymore (see https://github.com/whatwg/dom/pull/1152).

As for the above comment, the reason the tests was failing with the dependent signal approach, was largely due to dependent signals not forwarding one signals abort status to all other sibling signals. 

You can try this yourself with:

```js
const a = new AbortController();
const b = new AbortController();
const c = AbortSignal.any([a.signal, b.signal]);

a.signal.addEventListener('abort', () => console.log('a aborted')); // DOES NOT CALL
b.signal.addEventListener('abort', () => console.log('b aborted')); // called
c.addEventListener('abort', () => console.log('c aborted')); // called

b.abort();

console.log('a', a.signal.aborted); // false <~ aborting b does not abort a
console.log('b', b.signal.aborted); // true
console.log('c', c.aborted); // true
```

Which in our case would be that `a` is our internal abort signal, `b` the `SubscribeOptions.signal`, and `c` the [deprived signal point 3](https://wicg.github.io/observable/#dom-observable-foreach) tells us to create.

So for wpt tests that check that passed-in abort signals ran if we aborted, is not how dependent signals work.

As for why the algorithm wouldn't run (from my comment above), when sending `signal` as opposed to `dependantSignal` is still a mystery. But at least with the latest changes, the code reads a lot more sensibly, and passes the same tests — though isn't reflective of the spec.

We want all promise returning operators to have a predictable abort and promise timing order — so  I will work with @domenic, @domfarolino and @keithamus on making those spec changes to reflect this, and wpt test consistency.

Given this feature is still behind a feature flag, and the wpt tests pass on both Chrome and _now us_ — I am inclined to not let that spec change halt landing this PR.

Reply:
^ with all that being said, I also know very little C++ (kind of learning as im going) and might be entirely likely that the ordering is a symptom of AbortSignals being passed as references, vs pointers. Or something.

My point on timing order and inconsistent specs still stands.

Reply:
I understand this response is entirely late, and I apologize for this getting lost in my inbox, but I'd like to follow-up here to see if there is anything I can do, and to make sure I understand your original concern. In the first comment in this thread, you believed that the internal observer should be created with `signal` instead of `dependentSignal`, but you couldn't do this without failing some tests. But I noticed the PR landed with you passing in `signal`, and the tests in question passing: https://github.com/WebKit/WebKit/pull/32099/files#diff-f5ec8493ab9017c0d501d5725e74b79bbe13541fad819b65ee4569bbdaa05418R6.

Unfortunately there is no commit history in this PR for me to detect what you might have changed in between writing this thread and merging the PR to get it to work, so I'm curious — what did you change to fix it?

The only thing that the signal passed into the internal observer is used for is to get aborted in the case ["next" error case](https://wicg.github.io/observable/#ref-for-abortcontroller-signal-abort%E2%91%A1). The spec uses AbortController to achieve this since we shouldn't abort AbortSignals directly (although it seems that's fine in WebKit: https://bugs.webkit.org/show_bug.cgi?id=282069). So it shouldn't actually matter which signal you pass into the internal observer: `signal` or `dependentSignal`. In fact, it seems like `signal` doesn't even need to exist? I feel like you can just create `dependentSignal` directly based on `options.signal` — [`signal` itself](https://github.com/WebKit/WebKit/pull/32099/files#diff-3b52f4377c8ea8a505ebfc45f0d8bd7e6e5388a6eb762fa7dca8ac12405936e9R121) seems like it's just an empty pass-through, which isn't adding any direct value. Is that right? In that case, I wonder if you can get rid of it and just pass `dependentSignal` directly into the internal observer's `::create()` method.